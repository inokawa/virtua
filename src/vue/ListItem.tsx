/** @jsxImportSource vue */
import {
  ref,
  defineComponent,
  watch,
  StyleValue,
  PropType,
  VNode,
  NativeElements,
} from "vue";
import { ItemResizeObserver } from "../core/resizer";
import { isRTLDocument } from "../core/environment";

/**
 * @internal
 */
export const ListItem = /*#__PURE__*/ defineComponent({
  props: {
    _children: { type: Object as PropType<VNode>, required: true },
    _resizer: {
      type: Function as PropType<ItemResizeObserver>,
      required: true,
    },
    _index: { type: Number, required: true },
    _offset: { type: Number, required: true },
    _hide: { type: Boolean },
    _isHorizontal: { type: Boolean },
    _isSSR: { type: Boolean },
    _as: { type: String as PropType<keyof NativeElements>, required: true },
  },
  setup(props) {
    const elementRef = ref<HTMLDivElement>();

    // The index may be changed if elements are inserted to or removed from the start of props.children
    watch(
      () => elementRef.value && props._index,
      (_, __, onCleanup) => {
        onCleanup(props._resizer(elementRef.value!, props._index));
      },
      {
        flush: "post",
      }
    );

    return () => {
      const {
        _children: children,
        _offset: offset,
        _hide: hide,
        _isHorizontal: isHorizontal,
        _isSSR: isSSR,
        _as: Element,
      } = props;

      const style: StyleValue = {
        position: hide && isSSR ? undefined : "absolute",
        [isHorizontal ? "height" : "width"]: "100%",
        [isHorizontal ? "top" : "left"]: "0px",
        [isHorizontal ? (isRTLDocument() ? "right" : "left") : "top"]:
          offset + "px",
        visibility: !hide || isSSR ? "visible" : "hidden",
      };
      if (isHorizontal) {
        style.display = "flex";
      }

      return (
        <Element ref={elementRef} style={style}>
          {children}
        </Element>
      );
    };
  },
});
