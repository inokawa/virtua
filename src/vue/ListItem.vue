<script setup lang="ts">
import {
    ref,
    watch,
    NativeElements,
    computed,
    Ref,
} from "vue";
import {
    ItemResizeObserver,
    isRTLDocument,
    StateVersion,
    VirtualStore,
} from "../core/index.js";
import type { ItemProps } from "./utils.js";

const props = defineProps<{
    _stateVersion: number,
    _store: VirtualStore,
    _children: VNode,
    _resizer: ItemResizeObserver,
    _index: number,
    _isHorizontal: boolean,
    _isSSR: boolean;
    _as: keyof NativeElements,
    _itemProps: ReturnType<ItemProps>
}>();

const elementRef = ref<HTMLDivElement>();

const offset = computed(() =>
    props._stateVersion && props._store.$getItemOffset(props._index)
);
const hide = computed(() =>
    props._stateVersion && props._store.$isUnmeasuredItem(props._index)
);

// The index may be changed if elements are inserted to or removed from the start of props.children
watch(
    () => elementRef.value && props._index,
    (_, __, onCleanup) => {
        onCleanup(props._resizer(elementRef.value!, props._index));
    },
    {
        flush: "post",
    }
);
</script>

<template>
    <component :is="props._as" ref="elementRef" :style="{
        contain: 'layout style',
        position: hide && props._isSSR ? undefined : 'absolute',
        [props._isHorizontal ? 'height' : 'width']: '100%',
        [props._isHorizontal ? 'top' : 'left']: '0px',
        [props._isHorizontal ? (isRTLDocument() ? 'right' : 'left') : 'top']: offset + 'px',
        visibility: !hide || props._isSSR ? 'visible' : 'hidden',
        ...(props._itemProps?.style ?? {}),
        display: props._isHorizontal ? 'inline-flex' : undefined,
    }" v-bind="props._itemProps">
        <component :is="props._children" />
    </component>
</template>
