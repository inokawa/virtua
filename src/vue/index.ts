/**
 * @module vue
 */
export { default as VList } from "./VList.vue";
export type { VListHandle } from "./VList.vue";
export { default as Virtualizer } from "./Virtualizer.vue";
export type { VirtualizerHandle } from "./Virtualizer.vue";
export { default as WindowVirtualizer } from "./WindowVirtualizer.vue";
export type { WindowVirtualizerHandle } from "./WindowVirtualizer.vue";
