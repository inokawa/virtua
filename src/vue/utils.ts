import { CSSProperties, VNode } from "vue";
import { ItemsRange } from "../core";

/**
 * @internal
 */
export const getKey = (e: VNode, i: number): Exclude<VNode["key"], null> => {
  const key = e.key;
  return key != null ? key : "_" + i;
};

/**
 * @internal
 */
export const isSameRange = (prev: ItemsRange, next: ItemsRange): boolean => {
  return prev[0] === next[0] && prev[1] === next[1];
};

export type ItemProps = (payload: {
  item: any;
  index: number;
}) => { [key: string]: any; style?: CSSProperties; class?: string } | undefined;
