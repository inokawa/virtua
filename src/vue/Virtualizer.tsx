/** @jsxImportSource vue */
import {
  ref,
  onMounted,
  defineComponent,
  onUnmounted,
  VNode,
  watch,
  ComponentOptionsMixin,
  SlotsType,
  ComponentOptionsWithObjectProps,
  ComponentObjectPropsOptions,
  PropType,
  NativeElements,
} from "vue";
import {
  SCROLL_IDLE,
  UPDATE_SCROLL_EVENT,
  UPDATE_SCROLL_END_EVENT,
  UPDATE_VIRTUAL_STATE,
  getOverscanedRange,
  createVirtualStore,
  ACTION_ITEMS_LENGTH_CHANGE,
  getScrollSize,
  ACTION_START_OFFSET_CHANGE,
} from "../core/store";
import { createResizer } from "../core/resizer";
import { createScroller } from "../core/scroller";
import { ScrollToIndexOpts } from "../core/types";
import { ListItem } from "./ListItem";
import { getKey } from "./utils";
import { microtask } from "../core/utils";

export interface VirtualizerHandle {
  /**
   * Get current scrollTop or scrollLeft.
   */
  readonly scrollOffset: number;
  /**
   * Get current scrollHeight or scrollWidth.
   */
  readonly scrollSize: number;
  /**
   * Get current offsetHeight or offsetWidth.
   */
  readonly viewportSize: number;
  /**
   * Get item offset from start.
   * @param index index of item
   */
  getItemOffset(index: number): number;
  /**
   * Scroll to the item specified by index.
   * @param index index of item
   * @param opts options
   */
  scrollToIndex(index: number, opts?: ScrollToIndexOpts): void;
  /**
   * Scroll to the given offset.
   * @param offset offset from start
   */
  scrollTo(offset: number): void;
  /**
   * Scroll by the given offset.
   * @param offset offset from current position
   */
  scrollBy(offset: number): void;
}

const props = {
  /**
   * The data items rendered by this component.
   */
  data: { type: Array, required: true },
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan: { type: Number, default: 4 },
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize: Number,
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift: Boolean,
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal: Boolean,
  /**
   * If you put an element before virtualizer, you have to define its height with this prop.
   */
  startMargin: { type: Number, default: 0 },
  /**
   * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
   */
  ssrCount: Number,
  /**
   * Reference to the scrollable element. The default will get the parent element of virtualizer.
   */
  scrollRef: Object as PropType<HTMLElement>,
  /**
   * Component or element type for container element.
   * @defaultValue "div"
   */
  as: { type: String as PropType<keyof NativeElements>, default: "div" },
  /**
   * Component or element type for item element.
   * @defaultValue "div"
   */
  item: { type: String as PropType<keyof NativeElements>, default: "div" },
} satisfies ComponentObjectPropsOptions;

export const Virtualizer = /*#__PURE__*/ defineComponent({
  props: props,
  emits: ["scroll", "scrollEnd", "rangeChange"],
  setup(props, { emit, expose, slots }) {
    let isSSR = !!props.ssrCount;

    const isHorizontal = props.horizontal;
    const containerRef = ref<HTMLDivElement>();
    const store = createVirtualStore(
      props.data.length,
      props.itemSize ?? 40,
      props.ssrCount,
      undefined,
      !props.itemSize
    );
    const resizer = createResizer(store, isHorizontal);
    const scroller = createScroller(store, isHorizontal);

    const rerender = ref(store._getStateVersion());
    const unsubscribeStore = store._subscribe(UPDATE_VIRTUAL_STATE, () => {
      rerender.value = store._getStateVersion();
    });

    const unsubscribeOnScroll = store._subscribe(UPDATE_SCROLL_EVENT, () => {
      emit("scroll", store._getScrollOffset());
    });
    const unsubscribeOnScrollEnd = store._subscribe(
      UPDATE_SCROLL_END_EVENT,
      () => {
        emit("scrollEnd");
      }
    );

    onMounted(() => {
      isSSR = false;

      microtask(() => {
        const assignScrollableElement = (e: HTMLElement) => {
          resizer._observeRoot(e);
          scroller._observe(e);
        };
        if (props.scrollRef) {
          // parent's ref doesn't exist when onMounted is called
          assignScrollableElement(props.scrollRef!);
        } else {
          assignScrollableElement(containerRef.value!.parentElement!);
        }
      });
    });
    onUnmounted(() => {
      unsubscribeStore();
      unsubscribeOnScroll();
      unsubscribeOnScrollEnd();
      resizer._dispose();
      scroller._dispose();
    });

    watch(
      () => props.data.length,
      (count) => {
        store._update(ACTION_ITEMS_LENGTH_CHANGE, [count, props.shift]);
      }
    );
    watch(
      () => props.startMargin,
      (value) => {
        store._update(ACTION_START_OFFSET_CHANGE, value);
      },
      { immediate: true }
    );

    watch(
      [rerender, store._getJumpCount],
      ([, count], [, prevCount]) => {
        if (count === prevCount) return;

        scroller._fixScrollJump();
      },
      { flush: "post" }
    );

    watch(
      [rerender, store._getRange],
      ([, [start, end]], [, [prevStart, prevEnd]]) => {
        if (prevStart === start && prevEnd === end) return;

        emit("rangeChange", start, end);
      },
      { flush: "post" }
    );

    expose({
      get scrollOffset() {
        return store._getScrollOffset();
      },
      get scrollSize() {
        return getScrollSize(store);
      },
      get viewportSize() {
        return store._getViewportSize();
      },
      getItemOffset: store._getItemOffset,
      scrollToIndex: scroller._scrollToIndex,
      scrollTo: scroller._scrollTo,
      scrollBy: scroller._scrollBy,
    } satisfies VirtualizerHandle);

    return () => {
      rerender.value;

      const Element = props.as;
      const ItemElement = props.item;
      const count = props.data.length;

      const [startIndex, endIndex] = store._getRange();
      const scrollDirection = store._getScrollDirection();
      const totalSize = store._getTotalSize();

      const items: VNode[] = [];
      for (
        let [i, j] = getOverscanedRange(
          startIndex,
          endIndex,
          props.overscan,
          scrollDirection,
          count
        );
        i <= j;
        i++
      ) {
        const e = slots.default({ item: props.data![i]!, index: i })[0]!;
        items.push(
          <ListItem
            key={getKey(e, i)}
            _resizer={resizer._observeItem}
            _index={i}
            _offset={store._getItemOffset(i)}
            _hide={store._isUnmeasuredItem(i)}
            _children={e}
            _isHorizontal={isHorizontal}
            _isSSR={isSSR}
            _as={ItemElement}
          />
        );
      }

      return (
        <Element
          ref={containerRef}
          style={{
            // contain: "content",
            overflowAnchor: "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
            flex: "none", // flex style can break layout
            position: "relative",
            visibility: "hidden", // TODO replace with other optimization methods
            width: isHorizontal ? totalSize + "px" : "100%",
            height: isHorizontal ? "100%" : totalSize + "px",
            pointerEvents: scrollDirection !== SCROLL_IDLE ? "none" : undefined,
          }}
        >
          {items}
        </Element>
      );
    };
  },
} as ComponentOptionsWithObjectProps<
  typeof props,
  VirtualizerHandle,
  {},
  {},
  {},
  ComponentOptionsMixin,
  ComponentOptionsMixin,
  {
    /**
     * Callback invoked whenever scroll offset changes.
     * @param offset Current scrollTop or scrollLeft.
     */
    scroll: (offset: number) => void;
    /**
     * Callback invoked when scrolling stops.
     */
    scrollEnd: () => void;
    /**
     * Callback invoked when visible items range changes.
     */
    rangeChange: (
      /**
       * The start index of viewable items.
       */
      startIndex: number,
      /**
       * The end index of viewable items.
       */
      endIndex: number
    ) => void;
  },
  string,
  {},
  string,
  SlotsType<{ default: (arg: { item: any; index: number }) => VNode[] }>
>);
