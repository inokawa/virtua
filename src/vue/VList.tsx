/** @jsxImportSource vue */
import {
  defineComponent,
  ComponentOptionsMixin,
  SlotsType,
  ComponentOptionsWithObjectProps,
  ComponentObjectPropsOptions,
  ref,
  VNode,
} from "vue";
import { Virtualizer, VirtualizerHandle } from "./Virtualizer";

interface VListHandle extends VirtualizerHandle {}

const props = {
  /**
   * The data items rendered by this component.
   */
  data: { type: Array, required: true },
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan: { type: Number, default: 4 },
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize: Number,
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift: Boolean,
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal: Boolean,
  /**
   * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
   */
  ssrCount: Number,
} satisfies ComponentObjectPropsOptions;

export const VList = /*#__PURE__*/ defineComponent({
  props: props,
  emits: ["scroll", "scrollEnd", "rangeChange"],
  setup(props, { emit, expose, slots }) {
    const horizontal = props.horizontal;

    const onScroll = (offset: number) => {
      emit("scroll", offset);
    };
    const onScrollEnd = () => {
      emit("scrollEnd");
    };
    const onRangeChange = (start: number, end: number) => {
      emit("rangeChange", start, end);
    };

    const handle = ref<InstanceType<typeof Virtualizer>>();

    expose({
      get scrollOffset() {
        return handle.value!.scrollOffset;
      },
      get scrollSize() {
        return handle.value!.scrollSize;
      },
      get viewportSize() {
        return handle.value!.viewportSize;
      },
      getItemOffset: (...args) => handle.value!.getItemOffset(...args),
      scrollToIndex: (...args) => handle.value!.scrollToIndex(...args),
      scrollTo: (...args) => handle.value!.scrollTo(...args),
      scrollBy: (...args) => handle.value!.scrollBy(...args),
    } satisfies VListHandle);

    return () => {
      return (
        <div
          style={{
            display: horizontal ? "inline-block" : "block",
            [horizontal ? "overflowX" : "overflowY"]: "auto",
            contain: "strict",
            width: "100%",
            height: "100%",
          }}
        >
          <Virtualizer
            ref={handle}
            data={props.data}
            overscan={props.overscan}
            itemSize={props.itemSize}
            shift={props.shift}
            ssrCount={props.ssrCount}
            horizontal={horizontal}
            onScroll={onScroll}
            onScrollEnd={onScrollEnd}
            onRangeChange={onRangeChange}
          >
            {slots}
          </Virtualizer>
        </div>
      );
    };
  },
} as ComponentOptionsWithObjectProps<
  typeof props,
  VListHandle,
  {},
  {},
  {},
  ComponentOptionsMixin,
  ComponentOptionsMixin,
  {
    /**
     * Callback invoked whenever scroll offset changes.
     * @param offset Current scrollTop or scrollLeft.
     */
    scroll: (offset: number) => void;
    /**
     * Callback invoked when scrolling stops.
     */
    scrollEnd: () => void;
    /**
     * Callback invoked when visible items range changes.
     */
    rangeChange: (
      /**
       * The start index of viewable items.
       */
      startIndex: number,
      /**
       * The end index of viewable items.
       */
      endIndex: number
    ) => void;
  },
  string,
  {},
  string,
  SlotsType<{ default: (arg: { item: any; index: number }) => VNode[] }>
>);
