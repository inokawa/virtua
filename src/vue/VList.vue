<script setup lang="ts" generic="T">
import { ref, defineEmits, defineProps, defineExpose, VNode } from "vue";
import Virtualizer from "./Virtualizer.vue";
import type { VirtualizerHandle } from "./Virtualizer.vue";
import type { ItemProps } from "./utils.js";

export interface VListHandle extends VirtualizerHandle { }

interface VListProps<T> {
    /**
     * The data items rendered by this component.
     */
    data: T[];
    /**
     * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
     * @defaultValue 4
     */
    overscan?: number;
    /**
     * Item size hint for unmeasured items in pixels. It will help to reduce scroll jump when items are measured if used properly.
     *
     * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
     * - If set, you can opt out estimation and use the value as initial item size.
     */
    itemSize?: number;
    /**
     * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
     */
    shift?: boolean;
    /**
     * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
     */
    horizontal?: boolean;
    /**
     * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
     */
    ssrCount?: number;
    /**
     * A function that provides properties/attributes for item element
     *
     * **This prop will be merged into `item` prop in the future**
     */
    itemProps?: ItemProps;
    /**
     * List of indexes that should be always mounted, even when off screen.
     */
    keepMounted?: readonly number[];
}

defineSlots<{
    default: (props: { item: T, index: number }) => VNode;
}>();
const props = withDefaults(defineProps<VListProps<T>>(), { overscan: 4 });
const emit = defineEmits<{ scroll: [offset: number]; scrollEnd: [] }>();

const horizontal = !!props.horizontal;

const onScroll = (offset: number) => {
    emit("scroll", offset);
};
const onScrollEnd = () => {
    emit("scrollEnd");
};

const handle = ref<VirtualizerHandle | null>(null);

defineExpose<VListHandle>({
    get scrollOffset() {
        return handle.value!.scrollOffset
    },
    get scrollSize() {
        return handle.value!.scrollSize
    },
    get viewportSize() {
        return handle.value!.viewportSize
    },
    findStartIndex: (...args) => handle.value!.findStartIndex(...args),
    findEndIndex: (...args) => handle.value!.findEndIndex(...args),
    getItemOffset: (...args) => handle.value!.getItemOffset(...args),
    getItemSize: (...args) => handle.value!.getItemSize(...args),
    scrollToIndex: (...args) => handle.value!.scrollToIndex(...args),
    scrollTo: (...args) => handle.value!.scrollTo(...args),
    scrollBy: (...args) => handle.value!.scrollBy(...args),
});
</script>

<template>
    <div :style="{
        display: horizontal ? 'inline-block' : 'block',
        [horizontal ? 'overflowX' : 'overflowY']: 'auto',
        contain: 'strict',
        width: '100%',
        height: '100%',
    }">
        <Virtualizer ref="handle" :data="props.data" :overscan="props.overscan" :itemSize="props.itemSize"
            :itemProps="props.itemProps" :shift="props.shift" :ssrCount="props.ssrCount" :horizontal="horizontal"
            :keepMounted="props.keepMounted" @scroll="onScroll" @scrollEnd="onScrollEnd">
            <template v-slot:["default"]="slotProps">
                <slot :name="'default'" v-bind="slotProps" />
            </template>
        </Virtualizer>
    </div>
</template>
