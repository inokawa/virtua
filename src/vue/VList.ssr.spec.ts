/**
 * @vitest-environment node
 */
import { it, describe, expect } from "vitest";
import { createSSRApp } from "vue";
import { renderToString } from "vue/server-renderer";
import { VList } from "./VList";
import { JSDOM } from "jsdom";

const LIST_ID = "list-id";

describe("SSR", () => {
  it("should render items with renderToString and vertical", async () => {
    const COUNT = 10;
    const OVERSCAN = 4;
    const html = await renderToString(
      createSSRApp({
        components: { VList },
        template: `
        <VList :data="${JSON.stringify(
          Array.from({ length: 1000 }).map((_, i) => i)
        )}" :id="'${LIST_ID}'" :ssrCount="${COUNT}" :overscan="${OVERSCAN}" #default="{item}">
          <div :key="item">
            {{ item }}
          </div>
        </VList>
        `,
      })
    );
    expect(html).toMatchSnapshot();

    expect(
      new JSDOM(html).window.document.getElementById(LIST_ID)!.children[0]!
        .childElementCount
    ).toEqual(COUNT + OVERSCAN);
  });

  it("should render items with renderToString and horizontal", async () => {
    const COUNT = 10;
    const OVERSCAN = 4;
    const html = await renderToString(
      createSSRApp({
        components: { VList },
        template: `
        <VList :data="${JSON.stringify(
          Array.from({ length: 1000 }).map((_, i) => i)
        )}" :id="'${LIST_ID}'" :ssrCount="${COUNT}" :overscan="${OVERSCAN}" #default="{item}" :horizontal="true">
          <div :key="item">
            {{ item }}
          </div>
        </VList>
        `,
      })
    );
    expect(html).toMatchSnapshot();

    expect(
      new JSDOM(html).window.document.getElementById(LIST_ID)!.children[0]!
        .childElementCount
    ).toEqual(COUNT + OVERSCAN);
  });
});
