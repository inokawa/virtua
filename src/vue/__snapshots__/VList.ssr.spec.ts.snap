// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SSR > should render items with renderToString and horizontal 1`] = `"<div style="display:inline-block;overflow-x:auto;contain:strict;width:100%;height:100%;" id="list-id"><div style="overflow-anchor:none;flex:none;position:relative;visibility:hidden;width:40000px;height:100%;"><div style="height:100%;top:0px;left:0px;visibility:visible;display:flex;"><div>0</div></div><div style="height:100%;top:0px;left:40px;visibility:visible;display:flex;"><div>1</div></div><div style="height:100%;top:0px;left:80px;visibility:visible;display:flex;"><div>2</div></div><div style="height:100%;top:0px;left:120px;visibility:visible;display:flex;"><div>3</div></div><div style="height:100%;top:0px;left:160px;visibility:visible;display:flex;"><div>4</div></div><div style="height:100%;top:0px;left:200px;visibility:visible;display:flex;"><div>5</div></div><div style="height:100%;top:0px;left:240px;visibility:visible;display:flex;"><div>6</div></div><div style="height:100%;top:0px;left:280px;visibility:visible;display:flex;"><div>7</div></div><div style="height:100%;top:0px;left:320px;visibility:visible;display:flex;"><div>8</div></div><div style="height:100%;top:0px;left:360px;visibility:visible;display:flex;"><div>9</div></div><div style="height:100%;top:0px;left:400px;visibility:visible;display:flex;"><div>10</div></div><div style="height:100%;top:0px;left:440px;visibility:visible;display:flex;"><div>11</div></div><div style="height:100%;top:0px;left:480px;visibility:visible;display:flex;"><div>12</div></div><div style="height:100%;top:0px;left:520px;visibility:visible;display:flex;"><div>13</div></div></div></div>"`;

exports[`SSR > should render items with renderToString and vertical 1`] = `"<div style="display:block;overflow-y:auto;contain:strict;width:100%;height:100%;" id="list-id"><div style="overflow-anchor:none;flex:none;position:relative;visibility:hidden;width:100%;height:40000px;"><div style="width:100%;left:0px;top:0px;visibility:visible;"><div>0</div></div><div style="width:100%;left:0px;top:40px;visibility:visible;"><div>1</div></div><div style="width:100%;left:0px;top:80px;visibility:visible;"><div>2</div></div><div style="width:100%;left:0px;top:120px;visibility:visible;"><div>3</div></div><div style="width:100%;left:0px;top:160px;visibility:visible;"><div>4</div></div><div style="width:100%;left:0px;top:200px;visibility:visible;"><div>5</div></div><div style="width:100%;left:0px;top:240px;visibility:visible;"><div>6</div></div><div style="width:100%;left:0px;top:280px;visibility:visible;"><div>7</div></div><div style="width:100%;left:0px;top:320px;visibility:visible;"><div>8</div></div><div style="width:100%;left:0px;top:360px;visibility:visible;"><div>9</div></div><div style="width:100%;left:0px;top:400px;visibility:visible;"><div>10</div></div><div style="width:100%;left:0px;top:440px;visibility:visible;"><div>11</div></div><div style="width:100%;left:0px;top:480px;visibility:visible;"><div>12</div></div><div style="width:100%;left:0px;top:520px;visibility:visible;"><div>13</div></div></div></div>"`;
